core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java

```diff
@@ -429,7 +430,7 @@ protected RexNode removeCorrelationExpr(
         ImmutableSortedMap.of());
   }
 
-  public @Nullable Frame decorrelateRel(Sort rel) {
+  public @Nullable Frame decorrelateRel(Sort rel, boolean isCorVarDefined) {
     //
```

Comment:
Is there other way to handle the `isCorVarDefined` variables ? For example, keep it as a internal state while decorrelating.

Reply:
Thanks for the feedback @danny0405 .
I guess it could be possible to keep it in an internal state, but I believe it might be not that simple. For example (if I am not mistaken) if we deal with a Correlate within a Correlate, when the inner one is done, it should not set the state back to "non-correlate" because we are still in the context of the outer one.
What do you think @thomasrebele  ?

Reply:
Indeed, @rubenada's example needs to be taken care of if we would remove the isCorVarDefined. By passing it as an parameter we avoid such subtle problems. We could translate it to an attribute and wrap the code that changes the attribute in a try-finally block:

```
boolean prevCorVarDefined = isCorVarDefined
try {
    // ...
}
finally {
    isCorVarDefined = prevCorVarDefined;
}
```
I'm slightly leaning towards using a parameter, which looks a bit cleaner to me. Does anyone have another suggestions how to handle this variable?

Reply:
IMHO the solution proposed in the PR, even though it probably is not the most elegant one, is effective and arguably less error-prone than other alternatives, so I'd lean towards accepting it and merging the PR as it is. Especially if we consider that 1.27 release is around the corner and we should get this fixed before that.
Any thoughts @danny0405 ?

Reply:
I have no ideas yet, keep an recursive state in the decorrelator is really a complex work and needs careful design logic, we can merge this PR first and refactor it later on.

Reply:
I agree, I think the safest approach would be merging this PR to get it fixed for 1.27; and then considering a potential refactoring in a subsequent version.
So, if there are no objections, I'll merge this PR in the next 24/48 h